### PR TITLE
Kafka Realtime Sink

### PR DESCRIPTION
**Pull request by nitinmotgi**
Plugin that allows one to write to Kafka in realtime.

### CODE REVIEW

File: kafka-plugins/pom.xml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      <artifactId>cdap-etl-common</artifactId>
+      <version>${cdap.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.kafka</groupId>
+      <artifactId>kafka_2.10</artifactId>
+      <version>0.8.2.2</version>
+    </dependency>
+    <dependency>
+      <groupId>info.batey.kafka</groupId>
```
Comment (chtyim) [inline kafka-plugins/pom.xml]: What is this for? If test only, should be in test scope

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  private static final String ACKS_REQUIRED = "request.required.acks";
+
+  // Kafka properties
+  private final Properties props = new Properties();
+  
+  // Kafka producer configuration
+  private ProducerConfig config;
+  
+  // Kafka producer handle
+  private org.apache.kafka.clients.producer.KafkaProducer<String, String> producer;
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java:75]: Why full name?

File: kafka-plugins/pom.xml (hunk 1 above)
Comment (nitinmotgi) [inline kafka-plugins/pom.xml]: It's not required. Removed it.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 2 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java:75]: This is because the plugin class is also called KafkaProducer.

**Comment by nitinmotgi** [comment]
@chtyim please review and let me know. Thanks.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+@Name("KafkaProducer")
+@Description("Real-time Kafka producer")
+public class KafkaProducer extends RealtimeSink<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaProducer.class);
+
+  // Configuration for the plugin.
+  private final Config sconfig;
+  
+  // Static constants for configuring Kafka producer. 
+  private static final String BROKER_LIST = "bootstrap.servers";
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java:62]: Kafka producer supports quite a number of configuration. Shall we have an additional config map that can be provided by the user to specify any producer config?

http://kafka.apache.org/documentation.html#producerconfigs

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+ * data. It can also be configured to operate in sync or async mode.  
+ */
+@Plugin(type = "realtimesink")
+@Name("KafkaProducer")
+@Description("Real-time Kafka producer")
+public class KafkaProducer extends RealtimeSink<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaProducer.class);
+
+  // Configuration for the plugin.
+  private final Config sconfig;
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: What is "sconfig"? Better be just "config".

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+  private static final String KEY_SERIALIZER = "key.serializer";
+  private static final String VAL_SERIALIZER = "value.serializer";
+  private static final String CLIENT_ID = "client.id";
+  private static final String ACKS_REQUIRED = "request.required.acks";
+
+  // Kafka properties
+  private final Properties props = new Properties();
+  
+  // Kafka producer configuration
+  private ProducerConfig config;
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: this better call "producerConfig" or "kafkaConfig".

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 4 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: there is also kafka config. Two configs one for kafka and this is for sink.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+  // required for testing.
+  public KafkaProducer(Config config) {
+    this.sconfig = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    
+    if(!sconfig.isAsync.equalsIgnoreCase("true") && !sconfig.isAsync.equalsIgnoreCase("false")) {
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: Better use `Boolean.parseBoolean`.

Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: Also, missing space after `if`. Does the project configured with checkstyle?

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    int count = 0;
+
+    List<Schema.Field> fields = Lists.newArrayList();
+    
+    // For each object
+    for (StructuredRecord object : objects) {
+      
+      // Extract the field names from the object passed in. This is required
+      // because this information is not available in initialize or configuration phase.
+      if(!fieldsExtracted) {
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: Missing space after `if`.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 5 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: done

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        fields = object.getSchema().getFields();
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: Actually we don't need to memorize the `fields` with `fieldsExtracted`. Getting fields from schema is pretty much free.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+        
+        if (printer != null) {
+          printer.printRecord(objs);
+          body = writer.toString();
+        }
+      }
+      
+      // Message key.
+      String key = "no_key";
+      if(sconfig.key != null) {
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: missing space after `if`. There are quite a number of them. Make sure they get fixed and better have checkstyle configured.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+        key = object.get(sconfig.key);    
+      }
+      
+      // Extract the partition key from the record. If the partition key is 
+      // Integer then we use it as-is else
+      Integer partitionKey = 0;
+      if(sconfig.partitionField != null) {
+        if(object.get(sconfig.partitionField).getClass().isInstance(Integer.class)) {
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: `Integer.hashCode()` is the int value contained inside, hence no need to have special logic. Just do `object.get(partitionField).hashCode()` would do the trick.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+      
+      // Message key.
+      String key = "no_key";
+      if(sconfig.key != null) {
+        key = object.get(sconfig.key);    
+      }
+      
+      // Extract the partition key from the record. If the partition key is 
+      // Integer then we use it as-is else
+      Integer partitionKey = 0;
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: `int` instead of `Integer`.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 10 above)
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: The only thing you might want to guard against is the object doesn't have the partitionField, which will give you NPE if try to invoke `hashCode`.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+
+      // Write to all the configured topics
+      for(String topic : topics) {
+        partitionKey = partitionKey % producer.partitionsFor(topic).size();
+        if (isAsync) {
+          producer.send(new ProducerRecord<String, String>(topic, partitionKey, key, body), new Callback() {
+            @Override
+            public void onCompletion(RecordMetadata meta, Exception e) {
+              if (meta != null) {
+                context.getMetrics().count("kafka.async.success", 1);
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java:208]: This might lose metrics on stopping of the pipeline if that's acceptable.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+        context.getMetrics().count("kafka.producer.count", 1);
+      }
+    }
+    return count;
+  }
+  
+  @Override
+  public void destroy() {
+    super.destroy();
+    producer.close();
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: better do the close in a finally block.

**Comment by chtyim** [comment]
Did another pass.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 3 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java:62]: Understand. But, I want to keep it simple for now.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 7 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: done

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 8 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: done

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 9 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: No checkstyle yet added to this project. Will do in another PR.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 10 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: Done! & done!

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 11 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: done

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 13 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: done

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 12 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java:208]: yes. should be fine.

**Comment by nitinmotgi** [comment]
@chtyim addressed your comments.

**Comment by nitinmotgi** [comment]
Addressed your comments @chtyim

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+      String key = "no_key";
+      if (producerConfig.key != null) {
+        key = object.get(producerConfig.key);
+      }
+      
+      // Extract the partition key from the record. If the partition key is 
+      // Integer then we use it as-is else
+      int partitionKey = 0;
+      if (producerConfig.partitionField != null) {
+        if(object.get(producerConfig.partitionField) != null) {
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: missing space after `if`

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+
+      // Depending on the configuration create a body that needs to be 
+      // built and pushed to Kafka. 
+      String body = "";
+      if (producerConfig.format.equalsIgnoreCase("JSON")) {
+        body = StructuredRecordStringConverter.toJsonString(object);
+      } else {
+        // Extract all values from the structured record
+        List<Object> objs = Lists.newArrayList();
+        for(Schema.Field field : fields) {
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: space after for

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+    
+    @Name("brokers")
+    @Description("Specifies the connection string where Producer can find one or more brokers to " +
+      "determine the leader for each topic")
+    private String brokers;
+    
+    @Name("isasync")
+    @Description("Specifies whether an acknowledgment is required from broker that message was received. " +
+      "Default is FALSE")
+    private String isAsync;
```
Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: Why not use boolean type? Also, better call it "async" instead of "isAsync" (java bean naming convention).

Comment (chtyim) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: Also, should this field be optional (`@Nullable`) with a default of false?

**Comment by chtyim** [comment]
Just couple comments.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 14 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: thought I had fixed every where. I have double checked now. Thanks.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 15 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: fixed and made sure everywhere it's fine. Need to add checkstyle.

File: kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java (hunk 16 above)
Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: FE passes everything as string, so there is a problem there till it gets fixed.

Comment (nitinmotgi) [inline kafka-plugins/src/main/java/co/cask/hydrator/plugin/realtime/KafkaProducer.java]: changed isasync to async.

**Comment by nitinmotgi** [comment]
Addressed comments @chtyim

**Comment by chtyim** [comment]
Lgtm
